### ISSUE TITLE
Out of Memory on consolidate

### ISSUE THREAD

**Issue by mangecoeur** [TileDB-Inc/TileDB-Py#700]
I'm trying to consolidate fragments using tiledb.consolidate() on an array I constructed, but no matter what options I set the process eats all the memory (128GB) then gets killed by the OOM reaper. It reads very quickly some data off disc, then the disc read stops but the memory use grows and grows.

Example script I'm using (I also tried setting the memory budget first to a few GB then to a few MB but it didn't seem to do anything).

```python
config = tiledb.Config({"sm.consolidation.steps": 3, "sm.consolidation.mode": "fragments"})
ctx = tiledb.Ctx(config)
tiledb.consolidate(array_uri, ctx=ctx)
```

The array is about 8GB with the following schema:

```ArraySchema(
  domain=Domain(*[
    Dim(name='id_from', domain=(0, 2615281), tile='100', dtype='int32'),
    Dim(name='id_to', domain=(0, 2615281), tile='100', dtype='int32'),
  ]),
  attrs=[
    Attr(name='distance', dtype='float64', var=False, nullable=False),
  ],
  cell_order='row-major',
  tile_order='row-major',
  capacity=10000,
  sparse=True,
  allows_duplicates=False,
  coords_filters=FilterList([ByteShuffleFilter(),
       ZstdFilter(level=2)]),
)
```

I also occasionally get the same behaviour when reading the array - mostly I can slice it fine but sometimes a read triggers a very large and occasionally endless memory consumption growth.

**Comment by ihnorton** [comment]
Hi @mangecoeur, thanks for opening the issue. Is it possible to share any code or details about your write pattern? -- number of writes, sizes of writes, etc.? We will try to reproduce based on the schema, but any additional info you are able to share about the writes will be very helpful.

**Comment by ihnorton** [comment]
Hi @mangecoeur, I put together a quick script based on your schema, but wasn't able to reproduce so far -- hopefully we can tweak this a bit based on your write patterns in order to repro:

```
from tiledb import *
import numpy as np
import os, shutil

schema = ArraySchema(domain=Domain(*[
    Dim(name='id_from', domain=(0, 2615281), tile='100', dtype='int32'),
    Dim(name='id_to', domain=(0, 2615281), tile='100', dtype='int32'),
  ]),
  attrs=[
    Attr(name='distance', dtype='float64', var=False, nullable=False),
  ],
  cell_order='row-major',
  tile_order='row-major',
  capacity=10000,
  sparse=True,
  allows_duplicates=False,
  coords_filters=FilterList([ByteShuffleFilter(),
       ZstdFilter(level=2)]),
)

def consolidate(array_uri):
    config = tiledb.Config({"sm.consolidation.steps": 3, "sm.consolidation.mode": "fragments"})
    ctx = tiledb.Ctx(config)
    tiledb.consolidate(array_uri, ctx=ctx)

def write_array(array_uri, nelem=1000):
    d1 = schema.domain.dim(0).domain
    d2 = schema.domain.dim(1).domain
    c1 = np.random.randint(*d1, nelem)
    c2 = np.random.randint(*d2, nelem)
    data = np.random.rand(nelem)
    with tiledb.open(array_uri, "w") as A:
        A[c1,c2] = data


uri = "/tmp/py700"

if os.path.isdir(uri):
    shutil.rmtree(uri)

tiledb.Array.create(uri, schema)

for _ in range(100):
    write_array(uri)

consolidate(uri)
```

**Comment by ihnorton** [comment]
Also, could you please share your TileDB-Py version (I was using 0.10.1), and also whether setting the consolidation mode to `fragment_meta` ([docs](https://docs.tiledb.com/main/solutions/tiledb-embedded/api-usage/consolidation-and-vacuuming#fragment-metadata)) makes a difference?

**Comment by mangecoeur** [comment]
Hi, the OS is Ubuntu 20.04, versions are:

tiledb                    2.4.0                he87e0bf_2    conda-forge
tiledb-py                 0.10.1           py39hb79d20e_0    conda-forge

Running consolidation in mode 'fragment_meta' worked fine and completed very quickly.

The array stores distances between pairs of buildings, up to a max distance between buildings, to form a very sparse adjacency matrix. I iterate over about 25000 non-overlapping building clusters and for each cluster write a set of cells in the array. I have a separate lookup array that maps the (arbitrary) building integer IDs to array positions, and lookup is sorted using PostGIS st_geohash so that (in theory) buildings that are physically close together end up close together in the output array. The write step looks like this:

```python
        # calculate `dist` distances between node pairs (i_s, j_s)
        dist  = ...
        with tiledb.open(index_array_file, 'r') as A:
            coords_i = A.multi_index[list(i_s)]['index']
            coords_j = A.multi_index[list(j_s)]['index']

        with tiledb.open(array_name, 'w') as A:
            A[coords_i, coords_j] = dist
```

**Comment by ihnorton** [comment]
Hi @mangecoeur, thanks for the details. I will try some more tests to see if I can reproduce, but it would additionally be helpful to know:
- how many cells in each write (on average)?
- how many individual writes? (corresponding to # TileDB fragments -- is this 25,000?)
- what is the on-disk size of your final array?

Also, regarding this point:

> lookup is sorted using PostGIS st_geohash so that (in theory) buildings that are physically close together end up close together in the output array

I wanted to mention that there is an [option](https://tiledb-inc-tiledb.readthedocs-hosted.com/projects/tiledb-py/en/stable/python-api.html#array-schema) for Hilbert ordering when creating the ArraySchema, which has been very helpful for performance in similar geospatial use-cases:

```
schema = ArraySchema(domain=Domain(*[
    Dim(name='id_from', domain=(0, 2615281), tile='100', dtype='int32'),
  ...
  cell_order='hilbert',
  ...
)
```

**Comment by mangecoeur** [comment]
- how many cells in each write (on average)?
It varies a lot between clusters, but the mean cluster is 100 buildings giving around 100000 distance pairs

- how many individual writes? (corresponding to # TileDB fragments -- is this 25,000?)

yes 25000, one write per cluster

- what is the on-disk size of your final array?

8.3GB (very impressed by this, previous solution with dense hdf files per cluster was nearly 100GB)


> Hilbert ordering

Cool interesting to see this, I will try to test on some smaller subsets. Just to clarify, hilbert ordering would need to be used together with the geohash?

**Comment by ihnorton** [comment]
Thanks, I will try some further experiments to reproduce tomorrow.

> Just to clarify, hilbert ordering would need to be used together with the geohash?

Not necessarily: if the dimensions are specified as spatial coordinates (for example, floats or doubles representing lat/lon), then those coordinates could be written directly rather than using the geohash. When TileDB writes the cells to disk, the cells will first be arranged by sorting along a Hilbert curve filling the dimension space. Subsequently, each "tile" written to disk will then be much more likely to contain nearby cells, compared to using default row/col-major ordering. TileDB tiles store the MBR of each tile, so that a query may be satisfied by reading only those tiles whose MBR intersects the query ranges. With Hilbert ordering enabled, localized spatial queries can often be satisfied with fewer tile requests than other ordering -- so the required i/o and related computation (decompression etc.) is significantly reduced. cc @normanb @stavrospapadopoulos for any further clarification/correction there 😄.

**Comment by stavrospapadopoulos** [comment]
Yeah, my take is that Hilbert should be used standalone, as long as we properly constrain the array domain, since our Hilbert solution bins each dimension in order to handle real numbers (apologies, all that is being documented as we speak and will be published soon on our docs).

**Comment by mangecoeur** [comment]
Thanks for the feedback! Following your advice, I modified the writes to produce larger writes (~100MB) instead of many small, and used hilbert ordering for the 2d array. This solved the consolidation problem. 

Additionally, for my reads, I realised I was using `multi_index` wrong - I had assumed it worked more like numpy advanced indexing, rather than producing the cross product of slices; as a result some of my reads were producing huge selections. Fixing both of these issues gives me fast and reliable reads 👍
